source/server/connection_handler_impl.h

```diff
@@ -88,19 +86,19 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler, NonCopyable {
      */
     void removeConnection(ActiveConnection& connection);
 
+    /**
+     * Create a new connection from a socket accepted by the listener.
+     */
+    void newConnection(Network::AcceptedSocketPtr&& socket);
+
     ConnectionHandlerImpl& parent_;
-    Network::FilterChainFactory& factory_;
     Network::ListenerPtr listener_;
     ListenerStats stats_;
+    std::list<ActiveSocketPtr> sockets_;
     std::list<ActiveConnectionPtr> connections_;
     const uint64_t listener_tag_;
-  };
-
-  struct SslActiveListener : public ActiveListener {
-    SslActiveListener(ConnectionHandlerImpl& parent, Ssl::ServerContext& ssl_ctx,
-                      Network::ListenSocket& socket, Network::FilterChainFactory& factory,
-                      Stats::Scope& scope, uint64_t listener_tag,
-                      const Network::ListenerOptions& listener_options);
+    Network::ListenerConfig& config_;
+    Filter::Listener::ProxyProtocol::ConfigSharedPtr legacy_stats_;
```

Comment:
Instead of doing it this way, can we do legacy by actually altering the configuration to create filters configs? That would unify all of the logic and be cleaner IMO.

Reply:
Will do this tomorrow.

Reply:
Should we remove the legacy flag support ('use_original_dst', 'use_proxy_protocol') from the v2 API, so do you want to keep them for convenience?

Reply:
I would probably do one of two things:
1) Don't document original DST / proxy proto as filters, but just instantiate them as such. (So no change to documentation for those filters).
2) Deprecate the options and then we can delete support for them in the 1.7.0 cycle.

I don't really have a strong opinion. @htuch or @alyssawilk any opinion? (The question is how to deal with original DST and proxy proto legacy config).

Reply:
One complication here is that in v2 lds config 'use_proxy_proto' flag is in the filter_chain rather than in the config root. I guess that this is done this way to make it possible to filter on the source address against a known ELB address and then select a filter_chain with proxy protocol, which would allow ELB and non-ELB traffic to be accepted by the same listener on the same port.

Filter chain matching is not yet implemented, so this issue is not apparent in the code as of now, but my design in this PR runs both proxy protocol and original dst filters in the same listener_filters list, which would be run to completion before the filter_chain matching is (to be) done. This is why the listener_filters are specified as part of the main listener config rather than a filter_chain config.

The reasoning for having the listener filters before the filter_chain matching is that in the case of use_original_dst, it is more meaningful to match on the destination address for filter_chain selection after the destination address has been restored. With regard to the destination address this same argument holds for proxy protocol, too. I had not thought about the source address filtering for proxy protocol before, but it seems it would be beneficial to be able to do filter chain matching both before and after running proxy protocol, first on the source address to figure out that you need to run proxy protocol, and then after to select the network filter chain based on the restored destination address.

I guess the most flexible work around for this would be to allow insertion of listener filters both at the listener and filter_chain level. Thoughts?

Reply:
@jrajahalme ACK, this seems more complicated than I thought. I'm investigating and will get back to you with some thoughts.

Reply:
Ugh. This is a mess. Thanks for pointing this out @jrajahalme. Honestly, I don't feel it's worth the extra complexity to support running the listener filters more than once. To me it's an edge case, unless other people feel really strongly about it.

If we want to support proxy proto at the filter chain level, I wonder if it could be a special case (not an exposed filter) that runs after the filter chain is matched? Otherwise, I'm kind of the opinion that we should move proxy_proto into the base listener and deprecate where it is now.

@htuch @ggreenway @lizan @PiotrSikora thoughts ^^^ ?

Reply:
> Deprecating both bool flags could take care of this problem as well? This is not an issue for v1, and it is not an issue for anyone using v2 API right now, as the filter chain matching is not implemented yet. I mean that maybe even that special casing for proxy_proto after filter chain match is not needed?

Yes, I would argue it is not explicitly needed (I can see why someone would ask for it, but I am skeptical of how often this would come about). I would be fine with deprecating both booleans and moving this into the base listener level only.

My concern really is that someone feels very strongly that we need to support the use case where we do proxy_proto depending on filter chain matching, so looking for feedback from the folks that I already mentioned.

Reply:
@jrajahalme I discussed with @PiotrSikora @lizan @ggreenway and we think that we should leave proxy proto on the filter chain, but we don't want to do extra matching after we run it. So basically, proxy proto should not be a normal filter that is documented and run as part of the listener filter chain.

I think internally, it's probably OK if this is a filter, we can use it as such and special case it later when we implement real filter chain matching. Here is what I would do concretely:

1) Remove proxy proto filter from docs. The fact that it is a filter is an implementation detail.
2) For now just install it as the last filter as you are doing now.
3) I would do the configuration conversion as we previously discussed, assuming a single filter chain match and just put in TODOs for fix it up later.

Does that make sense?